In [1]:
import pandas as pd

hr = pd.read_csv('hr_dataset_raw.csv')
bands = pd.read_csv('salary_bands.csv')
budget = pd.read_csv('budget_growth_h2_2025.csv')

hr # просмотр содержимого 1 дф

,month,employee_id,full_name,dept_l1,dept_l2,dept_l3,prof_group,role,grade,rate,salary,hire_date,manager_id
0,2025-05,1105,Матвей Крылов,Технологический блок,Инфраструктура,NaN,Инфраструктура,Старший специалист,15.0,1.0,123000,2022-03-01,NaN
1,2025-05,1039,Светлана Соколова,Технологический блок,Инфраструктура,NaN,Инфраструктура,Ведущий специалист,16.0,1.0,169000,2023-09-01,NaN
2,2025-01,1017,Полина Козлова,Технологический блок,аналитика данных,NaN,Аналитика данных,Младший специалист,13.0,1.0,146000,2021-11-20,NaN
3,2025-05,1075,Мария Морозова,Операционный блок,HR,NaN,HR,Стажёр,12.0,1.0,121000,2024-02-20,NaN
4,2025-05,1028,Максим Семёнов,Технологический блок,аналитика данных,NaN,Аналитика данных,Руководитель группы,17.0,0.3,56100,2021-11-20,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
739,2025-04,1088,Кирилл Морозов,Операционный блок,HR,NaN,HR,Ведущий специалист,16.0,"0,5",75000,2023-09-01,NaN
740,2025-06,1065,Елена Семёнова,Финансовый блок,Компенсации и льготы,NaN,Компенсации и льготы,Старший специалист,15.0,1.0,148000,2021-11-20,NaN
741,2025-01,1077,Мария Васильева,Операционный блок,HR,NaN,HR,Младший специалист,13.0,1.0,140000,2022-06-15,NaN
742,2025-03,1062,Фёдор Кузнецов,Финансовый блок,Компенсации и льготы,NaN,Компенсации и льготы,Специалист,14.0,100%,140000,05.11.2024,NaN


In [2]:
bands # просмотр содержимого 2 дф

,prof_group,grade,salary_min,salary_mid,salary_max
0,Разработка,13,153000,180000,207000
1,Разработка,14,162000,190000,218000
2,Разработка,15,170000,200000,230000
3,Разработка,16,178000,210000,241000
4,Разработка,17,187000,220000,253000
5,Разработка,18,196000,230000,264000
6,Аналитика данных,13,141000,166000,191000
7,Аналитика данных,14,149000,175000,201000
8,Аналитика данных,15,156000,184000,212000
9,Аналитика данных,16,164000,193000,222000


In [3]:
budget # просмотр содержимого 3 дф

,category,budget_growth_rub
0,Инженерия ПО,130000
1,Дата-аналитика,210000
2,Эксплуатация и DevOps,100000
3,Финансовое планирование,120000
4,C&B,220000
5,Персонал и найм,80000
6,Клиентский сервис,100000
7,ИТОГО,960000


In [4]:
# Структура 1 дф
hr.info()
print(hr.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 744 entries, 0 to 743
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   month        744 non-null    object 
 1   employee_id  744 non-null    int64  
 2   full_name    744 non-null    object 
 3   dept_l1      713 non-null    object 
 4   dept_l2      734 non-null    object 
 5   dept_l3      109 non-null    object 
 6   prof_group   744 non-null    object 
 7   role         744 non-null    object 
 8   grade        740 non-null    float64
 9   rate         744 non-null    object 
 10  salary       744 non-null    object 
 11  hire_date    744 non-null    object 
 12  manager_id   6 non-null      float64
dtypes: float64(2), int64(1), object(10)
memory usage: 75.7+ KB
       employee_id       grade  manager_id
count   744.000000  740.000000         6.0
mean   1060.854839   15.204054      1113.0
std      35.477446    5.572039         0.0
min    1000.000000   12.00000

После **hr.info()** заметно:

**rate** — тип object вместо float, необходимо посмотреть уникальные значения

**salary** — тип object вместо int, необходимо посмотреть уникальные значения

**grade** — тип float вместо int, и 4 пропуска (740 non-null из 744)

**dept_l3** — 109 заполненных из 744, столбец почти пустой

**manager_id** — 6 заполненных из 744, столбец почти пустой

**dept_l1** — 31 пропуск, при анализе оклада их не используем, данные не трогаем

После **describe()** заметно:

**grade** — максимум 118.0, скорее всего, опечатка, нужно исправить.

**manager_id** — 6 одинаковых значений (1113.0), столбец не несёт полезной информации, не используем.

In [5]:
# Структура 2 дф
bands.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   prof_group  34 non-null     object
 1   grade       34 non-null     int64 
 2   salary_min  34 non-null     int64 
 3   salary_mid  34 non-null     int64 
 4   salary_max  34 non-null     int64 
dtypes: int64(4), object(1)
memory usage: 1.5+ KB


**salary_bands** - 34 строки, 7 проф.групп, грейды от 12 до 18. Пропусков нет, типы корректные. Используем для сравнения текущих окладов с рыночными вилками.

In [6]:
print(bands.describe())

           grade     salary_min     salary_mid     salary_max
count  34.000000      34.000000      34.000000      34.000000
mean   15.000000  144441.176471  169911.764706  195352.941176
std     1.595448   24465.425420   28718.453203   32932.329472
min    12.000000   98000.000000  115000.000000  132000.000000
25%    14.000000  128250.000000  151250.000000  174250.000000
50%    15.000000  146000.000000  171500.000000  197000.000000
75%    16.000000  160750.000000  188750.000000  216750.000000
max    18.000000  196000.000000  230000.000000  264000.000000


Статистика по рыночным вилкам (salary_bands):

Минимальная нижняя граница вилки: 98 000 руб. (Поддержка, grade 13)

Максимальная верхняя граница вилки: 264 000 руб. (Разработка, grade 18)

Середина вилки: 169 912 руб. - ориентир для справедливого оклада

Будем сравнивать текущие оклады сотрудников с этими вилками чтобы выявить кто недоплачен.

In [7]:
# Структура 3 дф
budget.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 2 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   category           8 non-null      object
 1   budget_growth_rub  8 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 260.0+ bytes


budget - 8 строк, последняя строка «ИТОГО» (960 000 руб.) - это сумма всех категорий, в расчётах её не используем. Реальный бюджет распределён по 7 категориям.
Названия категорий в budget не совпадают с названиями проф.групп в hr (например, «Инженерия ПО» в budget = «Разработка» в hr), нужно будет вручную сопоставить категории.

In [8]:
# Сопоставляем категории бюджета с проф.группами из hr
budget_mapping = {'Инженерия ПО': 'Разработка',
    'Дата-аналитика': 'Аналитика данных',
    'Эксплуатация и DevOps': 'Инфраструктура',
    'Финансовое планирование': 'Бюджетирование',
    'C&B': 'Компенсации и льготы',
    'Персонал и найм': 'HR',
    'Клиентский сервис': 'Поддержка'}

print(pd.DataFrame(list(budget_mapping.items()), columns=['Категория бюджета', 'Проф. группа в hr']))

         Категория бюджета     Проф. группа в hr
0             Инженерия ПО            Разработка
1           Дата-аналитика      Аналитика данных
2    Эксплуатация и DevOps        Инфраструктура
3  Финансовое планирование        Бюджетирование
4                      C&B  Компенсации и льготы
5          Персонал и найм                    HR
6        Клиентский сервис             Поддержка


Названия категорий в таблице бюджета и названия проф.групп в hr написаны по-разному, но означают одно и то же (например, «Инженерия ПО» = «Разработка»). Таблица выше показывает кто кому соответствует - это понадобится когда будем распределять бюджет на повышения.

In [9]:
# Смотрим уникальные значения rate, salary и grade, чтобы понять, что не так с форматом
print(hr['rate'].unique())
print(hr['salary'].unique())
print(hr['grade'].unique())

['1.0' '0.3' '0.8' '1,0' '0.5' '0.75' '75%' '0,6' '0.6' '100%' '0,8' '30%'
 '0,5' '0,3' '60%' '80%' '0,75' '50%']
['123000' '169000' '146000' '121000' '56100' '148000' '158400' '176000'
 '113000' '132000' '230000' '140000' '42000' '210000' '247000' '84500'
 '228000' '207000' '188000' '160000' '178000' '190000' '185000' '39000'
 '191000' '110000' '45000' '75000' '135000 руб' '200000' '117000' '142000'
 '118 500' '118500' '92000' '129000' '224000' '108000' '174000' '151200'
 '110400' '109200' '71 000' '216000' '66500' '91800' '110 400' '95250'
 '179000' '93600' '202000' '189000' '135000' '253000' '51000' '194 000'
 '98400' '187000' '186000' '122000' '151 200' '103800' '231000' '85200'
 '76 000' '141000' '137000' '96600' '152000' '173000' '124800' '117 000'
 '183000' '184000' '121 000' '165000' '185000 руб' '90000' '139000'
 '113400' '151000' '99000' '85200 руб' '121800' '118500 руб' '92000 руб'
 '64 000' '175000' '64000' '142000 руб' '163000' '250000' '155000'
 '182000' '161000' '123 000

Что заметно после **unique()**:

**rate**: одно и то же значение записано в разных форматах — десятичная дробь с точкой (1.0), с запятой (1,0) и в процентах (100%). Для расчётов нужно привести всё к единому числовому формату от 0 до 1.

**salary**: пробелы внутри числа: 118 500, 71 000; слово «руб»: 135000 руб, 185000 руб; комбинация обоих: 85200 руб, 118500 руб. Необходимо это убрать и перевести в int.

**grade**: присутсвует опечатка в виде "118" и nan. Нужно убрать опечатку и посмотреть, что за недостающие значения

In [10]:
# Чистим salary: убираем слово "руб" и пробелы, переводим в число
hr['salary'] = hr['salary'].str.replace('руб', '', regex=False)
hr['salary'] = hr['salary'].str.replace(' ', '', regex=False)
hr['salary'] = hr['salary'].astype(int)

# Проверяем результат
print(hr['salary'].unique())

[123000 169000 146000 121000  56100 148000 158400 176000 113000 132000
 230000 140000  42000 210000 247000  84500 228000 207000 188000 160000
 178000 190000 185000  39000 191000 110000  45000  75000 135000 200000
 117000 142000 118500  92000 129000 224000 108000 174000 151200 110400
 109200  71000 216000  66500  91800  95250 179000  93600 202000 189000
 253000  51000 194000  98400 187000 186000 122000 103800 231000  85200
  76000 141000 137000  96600 152000 173000 124800 183000 184000 165000
  90000 139000 113400 151000  99000 121800  64000 175000 163000 250000
 155000 182000 161000 209000 106000 127000 144000  82500 222000 100000
 195000 145000 225000 167000 217000 116000 168000]


In [11]:
# Чистим rate: убираем % и заменяем запятые на точки, переводим в число
hr['rate'] = hr['rate'].str.replace('%', '', regex=False)
hr['rate'] = hr['rate'].str.replace(',', '.', regex=False)
hr['rate'] = hr['rate'].astype(float)

# Если значение больше 1, то это проценты, делим на 100
hr['rate'] = hr['rate'].apply(lambda x: x / 100 if x > 1 else x)

# Проверяем результат
print(hr['rate'].unique())

[1.   0.3  0.8  0.5  0.75 0.6 ]


In [12]:
# Находим строки с грейдом 118
print(hr[hr['grade'] == 118])

# Исправляем 118 на 18
hr['grade'] = hr['grade'].replace(118, 18)
print(hr['grade'].unique())

       month  employee_id   full_name               dept_l1         dept_l2  \
66   2025-04         1114  Глеб Фомин  Технологический блок  Инфраструктура   
571  2025-02         1114  Глеб Фомин  Технологический блок  Инфраструктура   

    dept_l3      prof_group             role  grade  rate  salary   hire_date  \
66      NaN  Инфраструктура  Ведущий инженер  118.0   1.0  202000  2022-01-01   
571     NaN  Инфраструктура  Ведущий инженер  118.0   1.0  202000  2022-01-01   

     manager_id  
66       1113.0  
571      1113.0  
[15. 16. 13. 12. 17. 14. 18. nan]


In [13]:
# Находим сотрудников с пустым грейдом
print(hr[hr['grade'].isna()])

# Смотрим вилки для их проф.групп чтобы определить грейд вручную
print(bands[bands['prof_group'].isin(['Разработка', 'Бюджетирование'])])

# Заполняем грейды вручную на основе salary_bands:
# Роман Иванов (разработка, зарплата 209000) — попадает в grade 16 (min 178000, mid 210000, max 241000)
# ближе к mid (отклонение 1000) чем grade 15 (mid 200000, отклонение 9000)
hr.loc[hr['employee_id'] == 1004, 'grade'] = 16

# Григорий Васильев (бюджетирование, зарплата 175000) — попадает в grade 15 (min 143000, mid 168000, max 193000)
hr.loc[hr['employee_id'] == 1050, 'grade'] = 15

# Проверка
print(hr['grade'].isna().sum())

       month  employee_id          full_name               dept_l1  \
159  2025-01         1004       Роман Иванов  Технологический блок   
381  2025-01         1004       Роман Иванов  Технологический блок   
530  2025-02         1050  Григорий Васильев                   NaN   
689  2025-02         1050  Григорий Васильев                   NaN   

            dept_l2   dept_l3      prof_group        role  grade  rate  \
159         Разраб.  Фронтенд      разработка  Специалист    NaN   1.0   
381         Разраб.  Фронтенд      разработка  Специалист    NaN   1.0   
530  Бюджетирование       NaN  Бюджетирование  Специалист    NaN   1.0   
689  Бюджетирование       NaN  Бюджетирование  Специалист    NaN   1.0   

     salary   hire_date  manager_id  
159  209000  05.11.2024         NaN  
381  209000  05.11.2024         NaN  
530  175000  2023-01-10         NaN  
689  175000  2023-01-10         NaN  
        prof_group  grade  salary_min  salary_mid  salary_max
0       Разработка     13 

In [14]:
# Переводим grade в int
hr['grade'] = hr['grade'].astype(int)
print(hr['grade'].unique())

[15 16 13 12 17 14 18]


In [15]:
# Проверим уник знач prof_group
print(hr['prof_group'].unique())

['Инфраструктура' 'Аналитика данных' 'HR' 'Компенсации и льготы'
 'Аналитика данных ' 'аналитика данных' 'Бюджетирование' 'Поддержка'
 'разработка' 'Разработка' 'Компенсации' 'Разработка ' 'Разраб.' 'C&B'
 'Компенсации и льготы ']


**prof_group** - значения одной группы записаны по-разному:

1) Разработка, разработка, Разраб., Разработка  (с пробелом)

2) Аналитика данных, аналитика данных, Аналитика данных  (с пробелом)

3) Компенсации и льготы, Компенсации и льготы  (с пробелом), Компенсации, C&B

Необходимо привести к единому стандарту.

In [16]:
# Приводим prof_group к единому формату, убираем пробелы по краям
hr['prof_group'] = hr['prof_group'].str.strip()

# Стандартизируем
hr['prof_group'] = hr['prof_group'].replace({
    'разработка': 'Разработка',
    'Разраб.': 'Разработка',
    'аналитика данных': 'Аналитика данных',
    'Компенсации': 'Компенсации и льготы',
    'C&B': 'Компенсации и льготы'
})

print(hr['prof_group'].unique())

['Инфраструктура' 'Аналитика данных' 'HR' 'Компенсации и льготы'
 'Бюджетирование' 'Поддержка' 'Разработка']


In [17]:
# Проверим уник знач hire_date
print(hr['hire_date'].unique())

['2022-03-01' '2023-09-01' '2021-11-20' '2024-02-20' '20.11.2021'
 '2022-06-15' '01.03.2022' '01.05.2019' '05.11.2024' '2020-01-15'
 '2021-07-01' '2023-01-10' '2021-03-01' '2023-06-01' '15.06.2022'
 '2024-11-05' '01.03.2021' '20.02.2024' '15.01.2020' '01.09.2023'
 '2022-01-01' '2022-04-01' '2020-09-01' '01.02.2023' '2025-05-01'
 '10.01.2023' '2019-05-01' '2022-08-01' '2023-07-01' '01.01.2022'
 '01.08.2022' '2023-02-01' '2023-05-01' '01.07.2023' '01.07.2021'
 '01.09.2020' '01.05.2023' '01.05.2025' '01.04.2022']


**hire_date** — два формата дат: 2022-03-01 (YYYY-MM-DD) и 01.03.2022 (DD.MM.YYYY), приводим к единому YYYY-MM-DD.

In [18]:
# Единый формат даты
hr['hire_date'] = pd.to_datetime(hr['hire_date'], format='mixed', dayfirst=True)
hr['hire_date'] = hr['hire_date'].dt.strftime('%Y-%m-%d')

print(hr['hire_date'].unique())

['2022-03-01' '2023-09-01' '2021-11-20' '2024-02-20' '2022-06-15'
 '2019-05-01' '2024-11-05' '2020-01-15' '2021-07-01' '2023-01-10'
 '2021-03-01' '2023-06-01' '2022-01-01' '2022-04-01' '2020-09-01'
 '2023-02-01' '2025-05-01' '2022-08-01' '2023-07-01' '2023-05-01']


In [19]:
# Проверка данных после чистки
hr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 744 entries, 0 to 743
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   month        744 non-null    object 
 1   employee_id  744 non-null    int64  
 2   full_name    744 non-null    object 
 3   dept_l1      713 non-null    object 
 4   dept_l2      734 non-null    object 
 5   dept_l3      109 non-null    object 
 6   prof_group   744 non-null    object 
 7   role         744 non-null    object 
 8   grade        744 non-null    int64  
 9   rate         744 non-null    float64
 10  salary       744 non-null    int64  
 11  hire_date    744 non-null    object 
 12  manager_id   6 non-null      float64
dtypes: float64(2), int64(3), object(8)
memory usage: 75.7+ KB


**Итог чистки**:

**salary** - убраны пробелы и «руб», тип int64

**rate** - приведён к единому формату 0–1, тип float64

**grade** - исправлена опечатка, заполнены 2 пропуска по salary_bands, тип int64

**prof_group** - стандартизированы 5 вариантов написания, 7 уникальных групп

**hire_date** - приведён к единому формату дат

**dept_l3**, **manager_id** - не используем в анализе (слишком много пропусков)

Итого: 122 сотрудника, данные готовы к анализу.

In [20]:
# Финальная статистика после чистки
print(hr[['salary', 'rate', 'grade']].describe())

              salary        rate       grade
count     744.000000  744.000000  744.000000
mean   148580.779570    0.894220   14.956989
std     49493.548122    0.200051    1.543721
min     39000.000000    0.300000   12.000000
25%    110400.000000    0.800000   14.000000
50%    151000.000000    1.000000   15.000000
75%    186000.000000    1.000000   16.000000
max    253000.000000    1.000000   18.000000


Статистика после чистки:

**salary**: средняя зарплата 148 581 руб., минимум 39 000 руб. (частичные ставки), максимум 253 000 руб.

**rate**: средняя ставка 0.89 - большинство сотрудников работают на полную ставку (медиана и 75 перцентиль = 1.0), есть сотрудники на частичной ставке (минимум 0.3)

**grade**: грейды от 12 до 18, средний грейд 15 - большинство сотрудников среднего уровня

In [21]:
hr.to_csv('hr_clean.csv', index=False)
print('Файл сохранён')

Файл сохранён


In [22]:
# Берём для каждого сотрудника только последний доступный месяц
hr_last = hr.sort_values('month').groupby('employee_id').last().reset_index()

print(hr_last['month'].value_counts())

month
2025-06    122
Name: count, dtype: int64


Для каждого сотрудника взят последний доступный месяц, все 122 сотрудника представлены в июне 2025.

In [26]:
hr_last.to_csv('hr_last.csv', index=False, decimal=',')
print('Готово')

Готово
